### PR TITLE
Fixes #9022: Add jinja2 templating support to ncf

### PR DESCRIPTION
**Pull request by amousset**
http://www.rudder-project.org/redmine/issues/9022

### CODE REVIEW

**Comment by amousset** [comment]
First passing test! (code is very incomplete though).

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 1]
```diff
@@ -0,0 +1,45 @@
+#####################################################################################
+# Copyright 2014 Normation SAS
```
Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: 2016

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#
+# @class_prefix file_from_template
+# @class_parameter destination
+
+bundle agent file_from_template_jinja2(source_template, destination)
+{
+
+  vars:
+      "jinja2_script" string => "/home/amousset/JINJA/cli.py";
+      "datatstate_file" string => "/tmp/last-datatstate.json";
```
Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Pretty sure this is not right :-)

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name File from a mustache template
```
Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: jinja2

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+# @description This is a bundle to build a file from a mustache template
+# @agent_version >=3.6
```
Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Really?

**Comment by amousset** [comment]
Not really needing a review, it is clearly not finished, just feedback on the general overview.

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+# @parameter destination     Destination file
+#
+# @class_prefix file_from_template
+# @class_parameter destination
+
+bundle agent file_from_template_jinja2(source_template, destination)
+{
+
+  vars:
+      "jinja2_script" string => "/home/amousset/JINJA/cli.py";
```
Comment (amousset) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: `10_ncf_internals/jinja2_templating.py`?

File: tree/30_generic_methods/file_from_template_jinja2.cf (hunk 2 above)
Comment (amousset) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: We do not seem to have a temporary directory for now in ncf, where could we put this? We could add the PID to the filename and remove it after to avoid breaking things.

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      "datatstate_file" string => "/tmp/last-datatstate.json";
+    state_dumped::
+      "content" string => execresult("/usr/bin/python ${jinja2_script} --strict ${source_template} ${datatstate_file}", "noshell");
```
Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Doesn't using a string for this mean we limit the content to 2k / 4k / 16k / whatever the current CFEngine limit is?

File: tree/30_generic_methods/file_from_template_jinja2.cf (hunk 5 above)
Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Hmmm. Since we added some python external dependencies in modules/ the other day, why not use that path? They will already be copied... But... wait... why do we need to copy them?

File: tree/30_generic_methods/file_from_template_jinja2.cf (hunk 2 above)
Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: No, that's not a good idea. Directories like "config.d" read *all* files in them, and will start getting duplicated configs, etc. We have use /var/rudder/tmp before.

Comment (amousset) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Yes, but the question is why do we get this in ncf?

Comment (jooooooon) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Hmm, sorry, somehow my answer doesn't seem to be related to your question. I must have read/typed too quickly.

Adding the PID to the filename is a great idea :-)

How about we use the `source_template` name + a suffix like `-$(this.promiser_pid).datastate`?

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+{
+
+  vars:
+      "promisers"         slist  => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "old_class_prefix" string => canonify("file_from_template_${destination}");
+      "class_prefix"      string => canonify(join("_", "promisers"));
+      "args"              slist  => { "${source_template}", "${destination}" };
+
+      "jinja2_script"   string => "${sys.workdir}/modules/templates/jinja2-templating.py";
+      "datatstate_file" string => "${source_template}.datatstate.${this.promiser_pid}.json";
```
Comment (ncharles) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: shouldn't it be datastate_file (no t in the middle) ?

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/file_from_template_jinja2.cf (hunk 7 above)
Comment (amousset) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Indeed!

**Comment by amousset** [comment]
Added tests for some failure cases.

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+      "promisers"         slist  => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "old_class_prefix" string => canonify("file_from_template_${destination}");
+      "class_prefix"      string => canonify(join("_", "promisers"));
+      "args"              slist  => { "${source_template}", "${destination}" };
+
+      "jinja2_script"   string => "${sys.workdir}/modules/templates/jinja2-templating.py";
+      "datastate_file" string => "${source_template}.datastate.${this.promiser_pid}.json";
+      "destination_tmp" string => "${source_template}.${this.promiser_pid}.new";
+    state_dumped::
+      # Here we put the output into the rrors variable. As the stdout is sent to a file, the output is only stderr,
```
Comment (peckpeck) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: *e*rror

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      "pass1" expression => "any";
+
+  files:
+    # Use a file copy to avoid any file size limitations of CFEngine variables.
+    pass2.templating_ok::
+      "${destination}"
+        copy_from         => local_dcp("${destination_tmp}"),
+        action            => immediate,
+        move_obstructions => "true",
+        classes           => classes_generic("${old_class_prefix}");
```
Comment (peckpeck) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: use classes_generic_two

File: tree/30_generic_methods/file_from_template_jinja2.cf [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+#
+# @class_prefix file_from_template
+# @class_parameter destination
+
+bundle agent file_from_template_jinja2(source_template, destination)
+{
+
+  vars:
+      "promisers"         slist  => { @{this.callers_promisers}, cf_null }, policy => "ifdefined";
+      "old_class_prefix" string => canonify("file_from_template_${destination}");
```
Comment (peckpeck) [inline tree/30_generic_methods/file_from_template_jinja2.cf]: Irregular indentation

File: tree/10_ncf_internals/modules/templates/jinja2-templating.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        sys.stderr.write(err)
+        sys.exit(1)
+
+    env = Environment(
+        loader=FileSystemLoader(os.path.dirname(template_path)),
+        keep_trailing_newline=True,
+    )
+
+    if opts.strict:
+      env.undefined = StrictUndefined
```
Comment (peckpeck) [inline tree/10_ncf_internals/modules/templates/jinja2-templating.py]: irregular indentation

File: tree/10_ncf_internals/modules/templates/jinja2-templating.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+import sys
+import os
+import sys
+from optparse import OptionParser
+
+import jinja2
+from jinja2 import Environment, FileSystemLoader, StrictUndefined
+
+try: import simplejson as json
+except ImportError: import json
```
Comment (peckpeck) [inline tree/10_ncf_internals/modules/templates/jinja2-templating.py]: I don't like single line try/except

File: tree/10_ncf_internals/modules/templates/jinja2-templating.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+def render(opts, args):
+    if len(args) == 1:
+        data = sys.stdin.read()
+    else:
+        path = os.path.join(os.getcwd(), os.path.expanduser(args[1]))
+        data = open(path).read()
```
Comment (peckpeck) [inline tree/10_ncf_internals/modules/templates/jinja2-templating.py]: You should close this somewhere

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified

File: tree/30_generic_methods/file_from_template_type.cf [hunk 14]
```diff
@@ -37,24 +37,26 @@ bundle agent file_from_template_type(source_template, destination, template_type
...
+    is_jinja2.template_exists::
+      "template" usebundle => file_from_template_jinja2("${source_template}", "${destination}");
 
     !template_exists::
-
       "template_absent" usebundle => _classes_failure("${old_class_prefix}");
       "template_absent" usebundle => _classes_failure("${class_prefix}");
 
-    any::
+    !is_jinja2::
```
Comment (peckpeck) [inline tree/30_generic_methods/file_from_template_type.cf]: !(is_jinja2.template_exists)

File: tree/30_generic_methods/file_from_template_type.cf [hunk 15]
```diff
@@ -37,24 +37,26 @@ bundle agent file_from_template_type(source_template, destination, template_type
...
       "sanitize" usebundle => _classes_sanitize("${class_prefix}");
       "sanitize" usebundle => _classes_sanitize("${old_class_prefix}");
       "report"   usebundle => _log("Build file ${destination} from ${template_type} template ${source_template}", "${old_class_prefix}", "${class_prefix}", @{args});
```
Comment (peckpeck) [inline tree/30_generic_methods/file_from_template_type.cf]: For new classes : 

is_jinja2.template_exists:
      "new result classes" usebundle => _classes_copy("${class_prefix}_XXX", "${class_prefix}");
      "report"             usebundle => _log("XXX", "", "${class_prefix}", @{args});

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Commit modified
